extensions/clusters/clusters.go

```diff
@@ -483,12 +492,58 @@ func WaitClusterToBeUpgraded(client *rancher.Client, clusterID string) (err erro
 	return
 }
 
-func isClusterInaccessible(messages []string) (isInaccessible bool) {
-	clusterCPErrorMessage := "Cluster health check failed: Failed to communicate with API server during namespace check" // For GKE
-	clusterModifiedErrorMessage := "the object has been modified"                                                        // For provisioning node driver K3s and RKE2
+// WaitOnClusterAfterSnapshot waits for a cluster to finish taking a snapshot and return to an active state.
+func WaitOnClusterAfterSnapshot(client *rancher.Client, clusterID string) error {
+	cluster, err := client.Steve.SteveType(ProvisioningSteveResourceType).ByID(clusterID)
+	if err != nil {
+		return err
+	}
+	isTransitioning := false
+
+	if cluster.State == nil {
+		isTransitioning = true
+	} else {
+		isTransitioning = cluster.State.Transitioning
+	}
+
+	if !isTransitioning {
+		err = kwait.PollUntilContextTimeout(context.TODO(), defaults.FiveHundredMillisecondTimeout, defaults.OneMinuteTimeout, true, func(ctx context.Context) (bool, error) {
+			cluster, err := client.Steve.SteveType(ProvisioningSteveResourceType).ByID(clusterID)
+			if err != nil {
+				return false, err
+			}
+
+			// note, this intentionally ignores cluster.State.Error, as that can sometimes appear during an upgrade during snapshots.
+			if cluster.State == nil {
+				return false, nil
+			}
+			return cluster.State.Transitioning, nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	err = kwait.PollUntilContextTimeout(context.TODO(), 1*time.Second, defaults.FifteenMinuteTimeout, true, func(ctx context.Context) (bool, error) {
```

Comment:
Do these states need to happen every time? If thats the case, can we consolidate both the waits? It was a bit harder to follow through and understand why each wait is necessary. If both waits always occur, maybe we could simplify the logic with a single polling something like:

        if cluster.State == nil {
            return false, nil
        }       

        if cluster.State.Transitioning {
            return cluster.State.Name == "active", nil
        }`

Reply:
I'm not sure I understand how this helps. 

if !isTransitioning block will wait for the cluster to enter a transitioning state, which we need because we don't want to throw a false positive.
the block outside the if condition is there to wait on the cluster status to return to active.

Reply:
the reason being to avoid multiple pollings is to simplify it. this is how the above one goes: it first waits for the cluster to enter the transitioning state, similar to the current implementation and once cluster transitioning begins, it continues polling until the cluster gets to the active state.

Reply:
I think the current statements make enough sense are are doing what we're expecting of the function.